### PR TITLE
Fix #1583: Add JDK8 java.lang.Integer methods.

### LINKED ISSUE #1583: Integer.parseUnsignedInt is not implemented
Makes it hard to parse 32-bit hex strings. Especially since Long parsing has a bug too (#1582).

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

**Comment by nicolasstucki** [comment]
Review by @sjrd

File: javalanglib/src/main/scala/java/lang/Integer.scala [hunk 1]
```diff
@@ -53,24 +53,32 @@ object Integer {
...
   final val MAX_VALUE = 2147483647
   final val SIZE = 32
+  final val BYTES = 4
 
   @inline def valueOf(intValue: scala.Int): Integer = new Integer(intValue)
   @inline def valueOf(s: String): Integer = valueOf(parseInt(s))
-
-  @inline def valueOf(s: String, radix: Int): Integer =
-    valueOf(parseInt(s, radix))
+  @inline def valueOf(s: String, radix: Int): Integer = valueOf(parseInt(s, radix))
```
Comment (sjrd) [inline javalanglib/src/main/scala/java/lang/Integer.scala]: This line is too long (> 80 characters).

File: javalanglib/src/main/scala/java/lang/Integer.scala [hunk 2]
```diff
@@ -53,24 +53,32 @@ object Integer {
...
+  def parseInt(s: String, radix: scala.Int): scala.Int =
+    parseIntImpl(s, radix, signed = true)
+
+  @inline def parseUnsignedInt(s: String): scala.Int = parseUnsignedInt(s, 10)
+
+  def parseUnsignedInt(s: String, radix: scala.Int): scala.Int =
+    parseIntImpl(s, radix, signed = false)
+
+  @inline
+  private def parseIntImpl(s: String, radix: scala.Int, signed: scala.Boolean): scala.Int = {
```
Comment (sjrd) [inline javalanglib/src/main/scala/java/lang/Integer.scala]: Line too long. 

Also, the callers of this method should be marked `@noinline`.

File: javalanglib/src/main/scala/java/lang/Integer.scala [hunk 3]
```diff
@@ -53,24 +53,32 @@ object Integer {
...
     def fail = throw new NumberFormatException(s"""For input string: "$s"""")
 
     if (s == null || s.size == 0 ||
-        radix < Character.MIN_RADIX ||
-        radix > Character.MAX_RADIX)
+      radix < Character.MIN_RADIX ||
+      radix > Character.MAX_RADIX)
```
Comment (sjrd) [inline javalanglib/src/main/scala/java/lang/Integer.scala]: Indentation issue.

File: javalanglib/src/main/scala/java/lang/Integer.scala [hunk 4]
```diff
@@ -81,20 +89,37 @@ object Integer {
         }
         val res = js.Dynamic.global.parseInt(s, radix).asInstanceOf[scala.Double]
 
-        if (res.isNaN || res > MAX_VALUE || res < MIN_VALUE)
+        if (res.isNaN ||
+          (signed && (res > MAX_VALUE || res < MIN_VALUE)) ||
+          (!signed && (res > 0xFFFFFFFFL || res < 0))) {
```
Comment (sjrd) [inline javalanglib/src/main/scala/java/lang/Integer.scala]: Same. Need 4 spaces.

Comment (sjrd) [inline javalanglib/src/main/scala/java/lang/Integer.scala]: Besides, the condition on 3 lines is too long anyway. I recommend putting the bounds checks in a local `@inline def`. 
Also consider writing the bounds checks based on an `if (signed) ... else ...`. The intent would be clearer.

File: javalanglib/src/main/scala/java/lang/Integer.scala [hunk 5]
```diff
@@ -81,20 +89,37 @@ object Integer {
...
+        }
       }
     }
   }
 
-  @inline def toString(i: scala.Int): String =
-    "" + i
+  @inline def toString(i: scala.Int): String = "" + i
+
+  @inline def toUnsignedString(i: Int, radix: Int): String = toStringBase(i, radix)
```
Comment (sjrd) [inline javalanglib/src/main/scala/java/lang/Integer.scala]: Too long ;-)

File: javalanglib/src/main/scala/java/lang/Integer.scala [hunk 6]
```diff
@@ -81,20 +89,37 @@ object Integer {
...
 
   @inline def compare(x: scala.Int, y: scala.Int): scala.Int =
     if (x == y) 0 else if (x < y) -1 else 1
 
+  @inline def compareUnsigned(x: scala.Int, y: scala.Int): scala.Int = {
+    val signBitDiff = (x >>> 31) - (y >>> 31)
+    if (signBitDiff != 0) signBitDiff
+    else compare(x, y)
```
Comment (sjrd) [inline javalanglib/src/main/scala/java/lang/Integer.scala]: Not the best implementation. This is better:
```scala
if (x == y) 0
else if (x.toUint > y.toUint) 1
else -1
```

File: javalanglib/src/main/scala/java/lang/Integer.scala [hunk 7]
```diff
@@ -81,20 +89,37 @@ object Integer {
...
+  }
+
+  @inline def toUnsignedLong(x: Int): scala.Long = {
+    import js.JSNumberOps._
+    x.toUint.toLong
```
Comment (sjrd) [inline javalanglib/src/main/scala/java/lang/Integer.scala]: Better implementation: `x.toLong & 0xffffffffL`

File: javalanglib/src/main/scala/java/lang/Integer.scala [hunk 8]
```diff
@@ -156,8 +191,19 @@ object Integer {
...
+  @inline def max(a: Int, b: Int): Int = Math.max(a, b)
+  @inline def min(a: Int, b: Int): Int = Math.min(a, b)
+
   @inline private[this] def toStringBase(i: scala.Int, base: scala.Int): String = {
     import js.JSNumberOps._
     i.toUint.toString(base)
   }
+
+  @inline private def asInt(n: scala.Double): scala.Int =
+    (n.asInstanceOf[js.Dynamic] | 0.asInstanceOf[js.Dynamic]).asInstanceOf[scala.Int]
```
Comment (sjrd) [inline javalanglib/src/main/scala/java/lang/Integer.scala]: Just write `Int`, not `scala.Int`.

File: test-suite/src/test/require-jdk8/org/scalajs/testsuite/javalib/IntegerTestOnJDK8.scala [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+\*                                                                      */
+package org.scalajs.testsuite.javalib
+
+import org.scalajs.jasminetest.JasmineTest
+
+object IntegerTestOnJDK8 extends JasmineTest {
+
+  // Explicitly define these as `var`'s to avoid any compile-time constant folding
+  var MaxValue: Int = Int.MaxValue
+  var MinValue: Int = Int.MinValue
```
Comment (sjrd) [inline test-suite/src/test/require-jdk8/org/scalajs/testsuite/javalib/IntegerTestOnJDK8.scala]: No need for `var`. Non final `val` will do the same.

File: test-suite/src/test/require-jdk8/org/scalajs/testsuite/javalib/IntegerTestOnJDK8.scala [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+
+  describe("java.lang.Integer on JDK8 or higher") {
+    it("Should be represented with logical 4 bytes") {
+      expect(Integer.BYTES).toEqual(4)
```
Comment (sjrd) [inline test-suite/src/test/require-jdk8/org/scalajs/testsuite/javalib/IntegerTestOnJDK8.scala]: This doesn't test anything, in fact. scalac will constant fold this value with that found on the JDK classpath anyway.

File: test-suite/src/test/require-jdk8/org/scalajs/testsuite/javalib/IntegerTestOnJDK8.scala [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+      expect(compare(0xFFFFFFFF, 0xFFFFFFFF)).toEqual(0)
+      expect(compare(0xEE6B2800, 0xFFFFFFFF)).toBeLessThan(0)
+      expect(compare(0xFFFFFFFF, 0xEE6B2800)).toBeGreaterThan(0)
+      expect(compare(0xEE6B2800, 3)).toBeGreaterThan(0)
+      expect(compare(3, 0xEE6B2800)).toBeLessThan(0)
+    }
+
+    it("should provide `toUnsignedLong`") {
+      def test(x: Int, y: Long): Unit =
+        expect(Integer.toUnsignedLong(x)).toEqual(y)
```
Comment (sjrd) [inline test-suite/src/test/require-jdk8/org/scalajs/testsuite/javalib/IntegerTestOnJDK8.scala]: Need `== y).toBeTruthy` when comparing longs.

**Comment by sjrd** [comment]
That's all.

File: test-suite/src/test/require-jdk8/org/scalajs/testsuite/javalib/IntegerTestOnJDK8.scala (hunk 9 above)
Comment (nicolasstucki) [inline test-suite/src/test/require-jdk8/org/scalajs/testsuite/javalib/IntegerTestOnJDK8.scala]: Took is from the original Integer test. I'll fix those too.

**Comment by nicolasstucki** [comment]
Updated.

File: javalanglib/src/main/scala/java/lang/Integer.scala [hunk 12]
```diff
@@ -53,16 +53,26 @@ object Integer {
   final val MIN_VALUE = -2147483648
   final val MAX_VALUE = 2147483647
   final val SIZE = 32
+  final val BYTES = 4
 
   @inline def valueOf(intValue: scala.Int): Integer = new Integer(intValue)
   @inline def valueOf(s: String): Integer = valueOf(parseInt(s))
-
   @inline def valueOf(s: String, radix: Int): Integer =
     valueOf(parseInt(s, radix))
```
Comment (sjrd) [inline javalanglib/src/main/scala/java/lang/Integer.scala]: The blank line should not disappear.

File: javalanglib/src/main/scala/java/lang/Integer.scala [hunk 13]
```diff
@@ -81,20 +91,40 @@ object Integer {
         }
         val res = js.Dynamic.global.parseInt(s, radix).asInstanceOf[scala.Double]
 
-        if (res.isNaN || res > MAX_VALUE || res < MIN_VALUE)
+        @inline def isOutOfBounds(): scala.Boolean = {
+          if(signed) res > MAX_VALUE || res < MIN_VALUE
```
Comment (sjrd) [inline javalanglib/src/main/scala/java/lang/Integer.scala]: Missing space after `if`.
Also, no need for the `()` in `isOutOfBounds()`, since this method does not have side effects.

**Comment by sjrd** [comment]
That's all.

**Comment by nicolasstucki** [comment]
Updated.

File: javalanglib/src/main/scala/java/lang/Integer.scala [hunk 14]
```diff
@@ -53,16 +53,27 @@ object Integer {
   final val MIN_VALUE = -2147483648
   final val MAX_VALUE = 2147483647
   final val SIZE = 32
+  final val BYTES = 4
 
   @inline def valueOf(intValue: scala.Int): Integer = new Integer(intValue)
   @inline def valueOf(s: String): Integer = valueOf(parseInt(s))
-
+
```
Comment (sjrd) [inline javalanglib/src/main/scala/java/lang/Integer.scala]: uh oh. Spaces at EOL.

**Comment by sjrd** [comment]
LGTM after the space fix (which, in theory, won't pass scalastyle).

**Comment by sjrd** [comment]
The commit is erroneously attributed to me.

**Comment by nicolasstucki** [comment]
Updated author of commit

**Comment by sjrd** [comment]
LGTM
